Action buttons will now only update when needed instead of every Life().

phil235 opened the pull request:
- The action buttons now update their icon instantly.
- Fixes versions of pickup(),equipped() and dropped not calling the parent.
- Fixes drone not being able to remove a defib from their storage.
- You can now cycle the mime mask by clicking it in your hand.
- The action buttons for hardsuit and hooded suits now only appears when you're wearing the suit.
- Created two mob helper procs getBeltSlot() and getBackSlot().
- Created /datum/species/proc/on_species_loss() to handle stuff when our race change, currently only used by jelly and slime race to remove their exotic blood from our reagents and to remove slime people's action buttons.
- <b>Inserting items into a mob's chest no longer gives that item's action button to that mob (it was a bug).</b>
- Refactored the item's action system. Items can now hold multiple actions more easily.
- The "set internals" button of tank items now turn green when it's used as internals.
- <b>Removed research scanner from drones (since cyborgs don't have it, it's more consistent)</b>
- Removed the ignore_madkadjust mask var.
- The sechailer mask now has an adjust mask action button, so I removed the adjust verb that it was using.
- The item's action are now created on item/New() instead of trying to create it every time someone picks the item up.
- I split hud/action.dm, the datum/action stuff is now in the datum folder (/datum/action.dm), whereas the code for action buttons is kept in the hud folder under action_button.dm. Also I moved some /datum/action code that was in some files back into datum/action.dm where it belongs.

:cl: phil235
rscadd: The action button of gas tanks and jetpacks now turn green when you turn your internals on.
/:cl:

phil235 committed:
Action buttons will now only update when needed instead of every Life().
The action buttons now update their icon instantly.
Fixes versions of pickup(),equipped() and dropped not calling the parent.
Fixes drone not being able to remove a defib from their storage.
You can now cycle the mime mask by clicking it in your hand.
The action buttons for hardsuit and hooded suits now only appears when you're wearing the suit.
Created two mob helper procs getBeltSlot() and getBackSlot().
Created /datum/species/proc/on_species_loss() to handle stuff when our race change, currently only used by jelly and slime race to remove their exotic blood from our reagents and to remove slime people's action buttons.

phil235 commented on code/_onclick/hud/action.dm:
```diff
@@ -44,37 +50,28 @@
 /datum/action/proc/IsAvailable()
-	return Checks()
-
-/datum/action/proc/Checks()// returns 1 if all checks pass
```
I'm merging Checks() and IsAvailable(), they do the same thing.

phil235 commented on code/_onclick/hud/action.dm:
```diff
@@ -44,37 +50,28 @@
 			return 0
 	if(check_flags & AB_CHECK_STUNNED)
-		if(owner.stunned)
+		if(owner.stunned || owner.weakened)
```
if you can't use the button when stunned, you sure as hell can't use it when weakened.

phil235 commented on code/_onclick/hud/action.dm:
```diff
@@ -44,37 +50,28 @@
 		if(owner.lying)
 			return 0
-	if(check_flags & AB_CHECK_ALIVE)
+	if(check_flags & AB_CHECK_CONSCIOUS)
```
This checks if the mob is conscious, why call it check_alive then??

phil235 commented on code/_onclick/hud/action.dm:
```diff
@@ -90,8 +87,8 @@
 /obj/screen/movable/action_button
-	var/datum/action/owner
-	screen_loc = "WEST,NORTH"
+	var/datum/action/linked_action
```
renamed to linked_action since "owner" is confusing, especially since /datum/action has its own owner var that references a mob.

RemieRichards commented on code/game/objects/items.dm:
```diff
@@ -387,7 +388,20 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
 // note this isn't called during the initial dressing of a player
 /obj/item/proc/equipped(mob/user, slot)
-	return
+	if(action_button_name)
```
No, stop.
Stop supporting this system.

it needs to DIE.

optimumtact replied:
if checks?

RemieRichards replied:
@optimumtact no...
The whole "By default items have one action datum reference, a var to control that one reference's type and a string to decide its name" system.

Its shit.
Action Datums? 100%, Amazing, Greatness.
Action Strings? get_out%, die, you suck

phil235 replied:
Alright, I'll make a proper system.

phil235 commented on code/_onclick/hud/action.dm:
```diff
@@ -90,8 +87,8 @@
-	var/datum/action/owner
-	screen_loc = "WEST,NORTH"
+	var/datum/action/linked_action
+	screen_loc = null
```
Need screen_loc to start null, I don't remember why...

ChangelingRain commented:
Phil is based as fuck.

phil235 commented on code/_onclick/hud/movable_screen_objects.dm:
```diff
@@ -43,7 +43,7 @@
 		screen_loc = "[screen_loc_X[1]]:[pix_X],[screen_loc_Y[1]]:[pix_Y]"
 
-	moved = TRUE
+	moved = screen_loc
```
might as well use the var to remember the screen_loc position.

phil235 commented on code/game/gamemodes/handofgod/actions.dm:
```diff
@@ -19,3 +19,7 @@
 	owner << "You say: [msg]"
+
+/datum/action/innate/godspeak/Destroy()
+	god = null
```
clearing the reference to the god mob.

phil235 commented on code/game/gamemodes/handofgod/god.dm:
```diff
@@ -51,6 +51,10 @@
+	for(var/X in prophet_hats)
+		var/obj/item/clothing/head/helmet/plate/crusader/prophet/P = X
+		if(P.speak2god && P.speak2god.god == src)
+			qdel(P.speak2god)
```
When the god dies every prophet hat loses its speak action.

phil235 commented on code/game/gamemodes/miniantags/abduction/abduction_gear.dm:
```diff
@@ -29,20 +29,19 @@
+
+/obj/item/clothing/suit/armor/abductor/vest/item_action_slot_check(slot, mob/user)
+	if(slot == slot_wear_suit) //we only give the mob the ability to activate the vest if he's actually wearing it.
+		return 1
```
This proc is used so the action button of the vest only shows up when worn and not just in your hand.

phil235 commented on code/game/gamemodes/miniantags/morph/morph.dm:
```diff
@@ -169,9 +169,6 @@
 
-/mob/living/simple_animal/hostile/morph/update_action_buttons() //So all eaten objects are not counted every life
-	return
-
```
no longer needed, we don't check inside the mob for items' action button.

phil235 commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -70,35 +70,40 @@
-	if(ishuman(src.loc))
-		var/mob/living/carbon/human/H = src.loc
+	if(ismob(src.loc))
+		var/mob/M = src.loc
```
so drones can remove the defib from their storage.

phil235 commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -172,16 +177,22 @@
 
+/obj/item/weapon/defibrillator/item_action_slot_check(slot, mob/user)
+	if(slot == user.getBackSlot())
+		return 1
```
Using getBackSlot means drones can actually deploy the paddles.

phil235 commented on code/game/objects/items/weapons/implants/implant_freedom.dm:
```diff
@@ -8,14 +8,13 @@
 		var/mob/living/carbon/C_imp_in = imp_in
 		C_imp_in.uncuff()
+	if(!uses)
+		qdel(src)
```
When the implant is out of use we delete it which removes the action button from the mob.

phil235 commented on code/game/objects/items/weapons/storage/storage.dm:
```diff
@@ -374,10 +374,6 @@
-
-/obj/item/weapon/storage/dropped(mob/user)
-	return
-
```
Why was this here? Nobody knows.

phil235 commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -1,50 +1,41 @@
 	throw_speed = 1
 	throw_range = 4
-
+	action_button_name = "Set Internals"
```
the tank's action to set internals now use item/var/action instead of creating a new variable.

phil235 commented on code/game/objects/items/weapons/tanks/tanks.dm:
```diff
@@ -1,50 +1,41 @@
+		H << "<span class='warning'>You need a mask!</span>"
+		return
+	if(H.wear_mask.mask_adjusted)
+		H.wear_mask.adjustmask(H)
```
when using the internals icon (the one above the health hud) the code tries to adjust the mask back up before trying to turn on internals, but the tank's action button wasn't doing it. This fixes that.

phil235 commented on code/game/objects/items/weapons/tanks/watertank.dm:
```diff
@@ -75,22 +80,21 @@
 	..()
 
 /obj/item/weapon/watertank/MouseDrop(obj/over_object)
-	var/mob/H = src.loc
```
Let's make a mob, and...let's call it H!

phil235 commented on code/modules/mob/living/life.dm:
```diff
@@ -49,10 +49,6 @@
-	handle_actions()
-
-	handle_regular_hud_updates()
-
```
The whole reason for this PR. Removing two lines.

phil235 commented on code/modules/mob/living/silicon/robot/life.dm:
```diff
@@ -30,7 +30,7 @@
 	diag_hud_set_borgcell()
 
-/mob/living/silicon/robot/handle_regular_hud_updates()
+/mob/living/silicon/robot/proc/handle_robot_hud_updates()
```
this proc is the next on my list of shit to remove from Life()

phil235 commented on code/modules/mob/living/simple_animal/slime/death.dm:
```diff
@@ -9,8 +9,6 @@
 			is_adult = 0
 			maxHealth = 150
-			var/datum/action/innate/slime/evolve/E = new
-			E.Grant(src)
```
crap, that's a mistake, will fix.

Core0verload commented:
>Inserting items into a mob's chest no longer gives that item's action button to that mob (it was a bug).

NO IT WASNT. It was totally a feature!

duncathan commented:
I disagree with removing that bug
It's been around long enough to call it a feature

Iamgoofball commented:
do I need to start implanting maglites in my chest and then shadowling hunting for you guys to stop complaining

duncathan commented:
That's fucking awesome tho
I want to enable that it's cool

duncathan commented:
Anyway if you want to remove it that's fine but I disagree with it being in what's otherwise objectively a good pr

phil235 committed:
Refactored the item's action system. Items can now hold multiple actions.
The "set internals" button of tank items now turn green when it's used as internals.
Removed research scanner from drones (since cyborgs don't have it, it's more consistent)
Removed the ignore_madkadjust mask var.
The sechailer mask now has an adjust mask action button, so I removed the adjust verb that it was using.
The item's action are now created on item/New() instead of trying to create it every time someone picks the item up.
I split hud/action.dm, the datum/action stuff is now in the datum folder (/datum/action.dm), whereas the code for action buttons is kept in the hud folder under action_button.dm. Also I moved some /datum/action code that was in some files back into datum/action.dm where it belongs.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into ActionButtonUpdateRefactor

Conflicts:
	code/modules/clothing/clothing.dm
	code/modules/mob/living/carbon/inventory.dm
	code/modules/mob/living/carbon/life.dm
	icons/mob/actions.dmi

phil235 committed:
Fixes error in mergeconflict fixing.

phil235 commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -106,54 +106,59 @@
-						C << "<span class='warning'>You don't have an oxygen tank!</span>"
+				C << "<span class='warning'>You don't have an oxygen tank!</span>"
+				return
+	C.update_action_buttons_icon()
```
this is the only real change in this file, we update the action buttons icon (since the internal tank button turns green when used as internals now).

phil235 commented on code/datums/action.dm:
```diff
@@ -0,0 +1,357 @@
+	name = "Toggle Mode"
+
+/datum/action/item_action/toggle_barrier_spread
+	name = "Toggle Barrier Spread"
```
Most of these are just used to set the appropriate name for the action button.

phil235 commented on code/datums/action.dm:
```diff
@@ -0,0 +1,357 @@
+		if(iscarbon(owner))
+			var/mob/living/carbon/C = owner
+			if(target == C.internal)
+				button.icon_state = "bg_internals_on"
```
background turns green when tank is used as internals.

phil235 commented on code/datums/action.dm:
```diff
@@ -0,0 +1,357 @@
+/datum/action/item_action/toggle_research_scanner/ApplyIcon(obj/screen/movable/action_button/current_button)
+	if(button_icon && button_icon_state)
+		var/image/img = image(button_icon, current_button, "scan_mode")
+		current_button.overlays += img
```
I changed the research scanner action to be an item action instead of a weird datum/action/innate.

phil235 commented on code/game/gamemodes/blob/powers.dm:
```diff
@@ -248,14 +248,6 @@
-	var/datum/game_mode/blob/B = ticker.mode
-	if(!owner.mind || !(owner.mind in B.infected_crew))
-		return 1
-	return 0
```
CheckRemoval() proc is completely removed, we no longer need to check this every tick.

phil235 commented on code/game/objects/items.dm:
```diff
@@ -33,11 +33,8 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
-	var/action_button_internal = 0 //If 1, bypass the inside check action buttons test for
-	var/datum/action/item_action/action = null
+	var/list/actions = list() //list of /datum/action's that this item has.
+	var/list/actions_types = list() //list of paths of action datums to give to the item on New().
```
less variables for /obj/item.

AnturK replied:
:+1: Never got to doing this.

RemieRichards replied:
@phil235 you could actually just use the same list.
(removing, newing, adding back to the list)

phil235 commented on code/datums/action.dm:
```diff
@@ -0,0 +1,357 @@
+	button.moved = FALSE //so the button appears in its normal position when given to another owner.
+	L.actions -= src
+	L.update_action_buttons()
+	owner = null
```
We no longer delete and recreate the action's button every time it's removed and granted again.

phil235 commented on code/modules/clothing/clothing.dm:
```diff
@@ -116,7 +116,6 @@ BLIND     // can't see anything
 	strip_delay = 40
 	put_on_delay = 40
 	var/mask_adjusted = 0
-	var/ignore_maskadjust = 1
```
this var was originally added because one single mask needed it...

phil235 commented on code/modules/clothing/glasses/engine_goggles.dm:
```diff
@@ -30,11 +27,15 @@
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(H.glasses == src)
+			H.update_sight()
```
the check is ishuman instead of iscarbon, since only humans can wear glasses.

phil235 commented on code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm:
```diff
@@ -5,34 +5,6 @@ These are general powers. Specific powers are stored under the appropriate alien
-		return 1
-
-	return 0
-
```
moved to datum/action.dm

phil235 commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -80,7 +80,6 @@
 		equip_to_slot_or_del(I, slot_head)
 
 	access_card.flags |= NODROP
-	scanner.Grant(src)
```
drones lose their scanner. (consistency with robots not having it, and also makes things easier codewise)

phil235 commented on code/modules/mob/mob_defines.dm:
```diff
@@ -84,7 +84,6 @@
 	var/datum/hud/hud_used = null
 
 	var/research_scanner = 0 //For research scanner equipped mobs. Enable to show research data when examining.
-	var/datum/action/innate/scan_mode/scanner = new
```
Ughhhhh....

phil235 commented on code/game/objects/items.dm:
```diff
@@ -131,10 +128,17 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
+/obj/item/New()
+	..()
+	for(var/path in actions_types)
+		new path(src)
```
We create the item's actions in New() by going through the actions_types list.

phil235 commented:
Ready for review

Core0verload commented:
@phil235, how about adding implanted item action button back?

RemieRichards commented on code/datums/action.dm:
```diff
@@ -0,0 +1,357 @@
+		return 0
+	if(target)
+		var/obj/item/I = target
+		I.ui_action_click(owner, src.type)
```
next goal would be to destroy ui_action_click(), but that can be done at another time.

RemieRichards commented on code/datums/action.dm:
```diff
@@ -0,0 +1,357 @@
+
+/datum/action/item_action/organ_action/New(Target)
+	..()
+	name = "Toggle [target.name]"
```
All organ actions are now "Toggle [organname]"? Why?

phil235 replied:
Because the current button names used by organ actions are all pretty much "Toggle [theorganname]" , check the old "organ_action_name = "stuff" "

RemieRichards replied:
Yes except you CANT give it a new name now.

it should be ```if(!name) name = "Toggle [target.name]"```

RemieRichards commented:
@phil235 Besides what I've commented, this is good.

phil235 commented:
>how about adding implanted item action button back?

Most actions shouldn't be accessible when inside your chest. Especially flashlights which are used to powergame against shadowlings (plus emitting light through your chest is not realistic at all). I haven't seen any item action that was worth keeping when inserted in you.

phil235 committed:
Fixes organ_action, they can now be named something other than "toggle [target.name]"

Core0verload commented:
>Most actions shouldn't be accessible when inside your chest. Especially flashlights which are used to powergame against shadowlings (plus emitting light through your chest is not realistic at all)

Do you really think a surgeon of SS13 can't implant the flashlight while leaving a lightbulb outside and "on" switch accessible? Do you really thing the concept of "realism" justifies removing fun features?

I'll bring action button for implanted items back myself, but with new system I don't even know how.

optimumtact commented:
people can still do that, it just has to be explicitly coded instead of just being a bug that granted people action buttons to stuff that was in their contents.
